<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cell type deconvolution and differential analysis â€¢ SCdeconR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Cell type deconvolution and differential analysis">
<meta property="og:description" content="SCdeconR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCdeconR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Cell_type_deconvolution_and_differential_analysis.html">Cell type deconvolution and differential analysis</a>
    </li>
    <li>
      <a href="../articles/Deconvolution_with_scaden_for_spatial_transcriptomics_data.html">Deconvolution with scaden for spatial transcriptomics data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Liuy12/SCdeconR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Cell type deconvolution and differential
analysis</h1>
            
            <h4 data-toc-skip class="date">Compiled: June 03, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Liuy12/SCdeconR/blob/HEAD/vignettes/Cell_type_deconvolution_and_differential_analysis.Rmd" class="external-link"><code>vignettes/Cell_type_deconvolution_and_differential_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>Cell_type_deconvolution_and_differential_analysis.Rmd</code></div>

    </div>

    
    
<p>For this tutorial, we will firstly construct an integrated reference
using single-cell RNA-seq (scRNA-seq) and single-nuclei RNA-seq
(snRNA-seq) data from normal breast tissue samples.</p>
<p>Then, we will simulate artificial bulk RNA-seq samples using the
constructed reference data. Predefined cell-type proportions will be
used to introduce heterogeneity between groups for the simulated bulk
samples.</p>
<p>We then use <code>OLS</code> method to deconvolute the bulk samples
and predict cell-type proportions for each sample. Differential
expression and pathway enrichment analysis will then be performed to
identify perturbed genes and pathways. We will also be examining the
effect of correcting for cell-type proportion differences on those
downstream analysis.</p>
<div class="section level2">
<h2 id="load-required-libraries">Load required libraries<a class="anchor" aria-label="anchor" href="#load-required-libraries"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Liuy12/SCdeconR/" class="external-link">SCdeconR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="co"># register backend for parallel processing</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://doFuture.futureverse.org" class="external-link">doFuture</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://doFuture.futureverse.org/reference/registerDoFuture.html" class="external-link">registerDoFuture</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multisession"</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="construct-integrated-reference-data">Construct integrated reference data<a class="anchor" aria-label="anchor" href="#construct-integrated-reference-data"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># path to reference datasets</span></span>
<span><span class="va">ref_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/refdata/sample1"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/refdata/sample2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># path to phenodata files</span></span>
<span><span class="va">phenodata_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/refdata/phenodata_sample1.txt"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/refdata/phenodata_sample2.txt"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># construct integrated reference using harmony algorithm</span></span>
<span><span class="va">refdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_ref.html">construct_ref</a></span><span class="op">(</span>ref_list <span class="op">=</span> <span class="va">ref_list</span>, phenodata_list <span class="op">=</span> <span class="va">phenodata_list</span>, data_type <span class="op">=</span> <span class="st">"cellranger"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"harmony"</span>, group_var <span class="op">=</span> <span class="st">"subjectid"</span>, nfeature_rna <span class="op">=</span> <span class="fl">50</span>, vars_to_regress <span class="op">=</span> <span class="st">"percent_mt"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">refdata</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 33734 features across 260 samples within 2 assays </span></span>
<span><span class="co">## Active assay: SCT (8465 features, 2118 variable features)</span></span>
<span><span class="co">##  1 other assay present: RNA</span></span>
<span><span class="co">##  4 dimensional reductions calculated: pca, harmony, tsne, umap</span></span></code></pre>
<p>Here we set <code>nfeature_rna</code> to be 50 due to limited number
of cells we included in the data. You might want to apply a higher
cutoff for your own reference data.</p>
<details><summary><strong>what does refdata looks like</strong>
</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># refdata is a seurat object with those metadata information.  use ?refdata for further</span></span>
<span><span class="co"># documentation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">refdata</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    260 obs. of  12 variables:</span></span>
<span><span class="co">##  $ orig.ident     : chr  "SeuratProject" "SeuratProject" "SeuratProject" "SeuratProject" ...</span></span>
<span><span class="co">##  $ nCount_RNA     : num  2093 1576 945 3684 20403 ...</span></span>
<span><span class="co">##  $ nFeature_RNA   : int  836 700 490 1341 3364 4632 1062 911 1767 447 ...</span></span>
<span><span class="co">##  $ cellid         : chr  "TAAGAGACACATGACT-1_1" "GACGTGCCAAGTAGTA-1_1" "TCGAGGCCACTTGGAT-1_1" "TTCTCAATCTTGGGTA-1_1" ...</span></span>
<span><span class="co">##  $ celltype       : chr  "Epithelial cells" "Epithelial cells" "Epithelial cells" "Epithelial cells" ...</span></span>
<span><span class="co">##  $ subjectid      : chr  "subject1" "subject1" "subject1" "subject1" ...</span></span>
<span><span class="co">##  $ cohort         : chr  "komentissuebank" "komentissuebank" "komentissuebank" "komentissuebank" ...</span></span>
<span><span class="co">##  $ percent_mt     : num  0.478 4.822 1.799 3.366 4.284 ...</span></span>
<span><span class="co">##  $ nCount_SCT     : num  759 887 746 627 574 558 553 627 626 744 ...</span></span>
<span><span class="co">##  $ nFeature_SCT   : int  446 589 446 292 144 176 286 311 376 395 ...</span></span>
<span><span class="co">##  $ SCT_snn_res.0.8: Factor w/ 6 levels "0","1","2","3",..: 5 5 5 5 5 5 5 5 1 5 ...</span></span>
<span><span class="co">##  $ seurat_clusters: Factor w/ 6 levels "0","1","2","3",..: 5 5 5 5 5 5 5 5 1 5 ...</span></span></code></pre>
</details>
</div>
<div class="section level2">
<h2 id="simulate-artificial-bulk-rna-seq-samples-based-on-integrated-reference">Simulate artificial bulk RNA-seq samples based on integrated
reference<a class="anchor" aria-label="anchor" href="#simulate-artificial-bulk-rna-seq-samples-based-on-integrated-reference"></a>
</h2>
<p>Ideally, you should split the integrated reference data into training
and testing, and use the training data to generate artificial bulk
samples. The testing data is used in deconvolution step. However, for
the sake of this tutorial, due to the small # of cells we included in
the <code>refdata</code>, we will not split the data into training and
testing.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create phenodata</span></span>
<span><span class="va">phenodata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>cellid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">refdata</span><span class="op">)</span>, celltypes <span class="op">=</span> <span class="va">refdata</span><span class="op">$</span><span class="va">celltype</span>, subjectid <span class="op">=</span> <span class="va">refdata</span><span class="op">$</span><span class="va">subjectid</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here is the number of cells per cell-type in refdata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">refdata</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##         Adipocyte Endothelial cells  Epithelial cells        Fibroblast </span></span>
<span><span class="co">##                20                40                40                40 </span></span>
<span><span class="co">##         Monocytes          NK cells         Pericytes            T cell </span></span>
<span><span class="co">##                40                20                40                20</span></span></code></pre>
<p>Next we generate two sets of bulk data that have different cell-type
proportions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># equal proportions across cell-types</span></span>
<span><span class="va">prop1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>celltypes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">refdata</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>, proportion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.125</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate 20 artificial bulk samples with the above cell-type proportions</span></span>
<span><span class="va">bulk_sim1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bulk_generator.html">bulk_generator</a></span><span class="op">(</span>ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">refdata</span>, slot <span class="op">=</span> <span class="st">"data"</span>, assay <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span>, phenodata <span class="op">=</span> <span class="va">phenodata</span>,</span>
<span>    num_mixtures <span class="op">=</span> <span class="fl">20</span>, prop <span class="op">=</span> <span class="va">prop1</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># high proportion of epithelial cells</span></span>
<span><span class="va">prop2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>celltypes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">refdata</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>, proportion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># generate 20 artificial bulk samples with the above cell-type proportions</span></span>
<span><span class="va">bulk_sim2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bulk_generator.html">bulk_generator</a></span><span class="op">(</span>ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">refdata</span>, slot <span class="op">=</span> <span class="st">"data"</span>, assay <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span>, phenodata <span class="op">=</span> <span class="va">phenodata</span>,</span>
<span>    num_mixtures <span class="op">=</span> <span class="fl">20</span>, prop <span class="op">=</span> <span class="va">prop2</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine the two datasets</span></span>
<span><span class="va">bulk_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">bulk_sim1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">bulk_sim2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">bulk_sim1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">bulk_sim2</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<details><summary><strong>what does simulated bulk data looks like</strong>
</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># bulk_generator returns a list of two elements.  the first element is simulated bulk RNA-seq</span></span>
<span><span class="co"># data, with rows representing genes, columns representing samples.  show first five samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">bulk_sim</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    8465 obs. of  5 variables:</span></span>
<span><span class="co">##  $ mix1_1: num  0.693 6.644 3.871 1.386 0.693 ...</span></span>
<span><span class="co">##  $ mix1_2: num  0.693 7.625 4.564 0.693 2.773 ...</span></span>
<span><span class="co">##  $ mix1_3: num  1.386 5.545 2.773 0.693 2.079 ...</span></span>
<span><span class="co">##  $ mix1_4: num  0.693 4.852 5.257 0 3.466 ...</span></span>
<span><span class="co">##  $ mix1_5: num  0.693 8.318 1.386 1.386 2.773 ...</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the second element is cell type proportions used to simulate the bulk RNA-seq data, with</span></span>
<span><span class="co"># rows representing cell types, columns representing samples.  show first five samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">bulk_sim</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    8 obs. of  5 variables:</span></span>
<span><span class="co">##  $ mix1_1: num  0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.125</span></span>
<span><span class="co">##  $ mix1_2: num  0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.125</span></span>
<span><span class="co">##  $ mix1_3: num  0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.125</span></span>
<span><span class="co">##  $ mix1_4: num  0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.125</span></span>
<span><span class="co">##  $ mix1_5: num  0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.125</span></span></code></pre>
</details>
</div>
<div class="section level2">
<h2 id="perform-cell-type-deconvolution-using-ols-algorithm">Perform cell-type deconvolution using OLS algorithm<a class="anchor" aria-label="anchor" href="#perform-cell-type-deconvolution-using-ols-algorithm"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># as mentioned ealier, we use the same reference data for deconvolution.  in practice, we</span></span>
<span><span class="co"># recommend to split your reference data into training and testing, and use testing data for</span></span>
<span><span class="co"># decovolution</span></span>
<span><span class="va">decon_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scdecon.html">scdecon</a></span><span class="op">(</span>bulk <span class="op">=</span> <span class="va">bulk_sim</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">refdata</span>, slot <span class="op">=</span> <span class="st">"data"</span>, assay <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span>,</span>
<span>    phenodata <span class="op">=</span> <span class="va">phenodata</span>, filter_ref <span class="op">=</span> <span class="cn">TRUE</span>, decon_method <span class="op">=</span> <span class="st">"OLS"</span>, norm_method_sc <span class="op">=</span> <span class="st">"LogNormalize"</span>,</span>
<span>    norm_method_bulk <span class="op">=</span> <span class="st">"TMM"</span>, trans_method_sc <span class="op">=</span> <span class="st">"none"</span>, trans_method_bulk <span class="op">=</span> <span class="st">"log2"</span>, marker_strategy <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></code></pre></div>
<details><summary><strong>what does deconvolution result looks like</strong>
</summary><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># scdecon returns a list of two elements the first element is a data.frame of predicted</span></span>
<span><span class="co"># cell-type proportions, with rows representing cell types, columns representing samples.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">decon_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  num [1:8, 1:40] 0.2121 0.1286 0.1319 0 0.0924 ...</span></span>
<span><span class="co">##  - attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   ..$ : chr [1:8] "Epithelial cells" "Monocytes" "Fibroblast" "T cell" ...</span></span>
<span><span class="co">##   ..$ : chr [1:40] "mix1_1" "mix1_2" "mix1_3" "mix1_4" ...</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the second element is a data.frame of fitting errors of the algorithm; first column</span></span>
<span><span class="co"># represents sample names, second column represents RMSEs.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">decon_res</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    40 obs. of  2 variables:</span></span>
<span><span class="co">##  $ sample: chr  "mix1_1" "mix1_2" "mix1_3" "mix1_4" ...</span></span>
<span><span class="co">##  $ RMSE  : num  7.14 7.1 6.48 7.78 6.69 ...</span></span></code></pre>
</details><p>We can then generate a bar plot of predicted cell proportions across
samples</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop_barplot.html">prop_barplot</a></span><span class="op">(</span>prop <span class="op">=</span> <span class="va">decon_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Cell_type_deconvolution_and_differential_analysis_files/figure-html/unnamed-chunk-12-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="perform-differential-expression-analysis">Perform differential expression analysis<a class="anchor" aria-label="anchor" href="#perform-differential-expression-analysis"></a>
</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare sampleinfo, group1 and group2 were the two bulk sets we simulated eariler.</span></span>
<span><span class="va">sampleinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group1"</span>, <span class="st">"group2"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare bulk samples</span></span>
<span><span class="va">bulk</span> <span class="op">&lt;-</span> <span class="va">bulk_sim</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># force data to be integers for DE purposes</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/mode.html" class="external-link">storage.mode</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"integer"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare predicted cell-type proportions</span></span>
<span><span class="va">prop</span> <span class="op">=</span> <span class="va">decon_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform DEA adjusting cell-type proportion differences</span></span>
<span><span class="va">deres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_de.html">run_de</a></span><span class="op">(</span>bulk <span class="op">=</span> <span class="va">bulk</span>, prop <span class="op">=</span> <span class="va">prop</span>, sampleinfo <span class="op">=</span> <span class="va">sampleinfo</span>, control <span class="op">=</span> <span class="st">"group1"</span>, case <span class="op">=</span> <span class="st">"group2"</span>,</span>
<span>    de_method <span class="op">=</span> <span class="st">"edgeR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform DEA without adjusting cell-type proportion differences</span></span>
<span><span class="va">deres_notadjust</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_de.html">run_de</a></span><span class="op">(</span>bulk <span class="op">=</span> <span class="va">bulk</span>, prop <span class="op">=</span> <span class="cn">NULL</span>, sampleinfo <span class="op">=</span> <span class="va">sampleinfo</span>, control <span class="op">=</span> <span class="st">"group1"</span>,</span>
<span>    case <span class="op">=</span> <span class="st">"group2"</span>, de_method <span class="op">=</span> <span class="st">"edgeR"</span><span class="op">)</span></span></code></pre></div>
<p>Next, we can compare the effect of adjusting for cell-type proportion
differences</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comparedeg_scatter.html">comparedeg_scatter</a></span><span class="op">(</span>results1 <span class="op">=</span> <span class="va">deres</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, results2 <span class="op">=</span> <span class="va">deres_notadjust</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, result_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"adjust for cell proportion"</span>,</span>
<span>    <span class="st">"not adjust for cell proportion"</span><span class="op">)</span>, fc_cutoff <span class="op">=</span> <span class="fl">1.5</span>, pval_cutoff <span class="op">=</span> <span class="fl">0.05</span>, pvalflag <span class="op">=</span> <span class="cn">TRUE</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Cell_type_deconvolution_and_differential_analysis_files/figure-html/unnamed-chunk-14-1.png" width="768" style="display: block; margin: auto;"></p>
<p>As you can see, many false positive differential genes were detected
without adjusting for cell proportion differences.</p>
</div>
<div class="section level2">
<h2 id="gene-set-enrichment-analysis">Gene-set enrichment analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis"></a>
</h2>
<p>We also provide several visualization options for pathway enrichment
results generated using <a href="https://www.gsea-msigdb.org/gsea/index.jsp" class="external-link">GSEA</a> software.
Note that those functions are not compatible with the <a href="https://github.com/GSEA-MSigDB/GSEA_R" class="external-link">R implementation</a> of
GSEA. Also remember to run <code>reformt_gmt()</code> to reformat the
gene-set names before using GSEA.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reformat_gmt.html">reformat_gmt</a></span><span class="op">(</span>gmtfile <span class="op">=</span> <span class="st">"/path/to/gmt/file/"</span>, outputfile <span class="op">=</span> <span class="st">"/path/to/reformatted/gmt/file/"</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## here is one example of reformatted gmt file for kegg pathway</span></span>
<span><span class="va">gmt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reformat_gmt.html">read_gmt</a></span><span class="op">(</span>gmtfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/gsea/gmtfile/kegg.gmt"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Prepare <code>.rnk</code> file.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prepare_rnk.html">prepare_rnk</a></span><span class="op">(</span>teststats <span class="op">=</span> <span class="va">deres</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, outputfile <span class="op">=</span> <span class="st">"/path/to/rnk/file/"</span><span class="op">)</span></span></code></pre></div>
<p>Use the reformatted <code>.gmt</code> file and <code>.rnk</code> file
to run GSEA. You can download GSEA <a href="https://www.gsea-msigdb.org/gsea/downloads.jsp" class="external-link">here</a>. Do not
use the R version implementation (our visualization functions are not
compatible with GSEA R implementation).</p>
<p>We included GSEA output using differential genes w/wo adjusting for
cell-type differences. We can now compare the differences between those
results.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comparegsea_scatter.html">comparegsea_scatter</a></span><span class="op">(</span>gseares_path1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/gsea/results/ct_adjust/"</span><span class="op">)</span>,</span>
<span>    gseares_path2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/gsea/results/ct_unadjust/"</span><span class="op">)</span>,</span>
<span>    result_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ct_adjust"</span>, <span class="st">"ct_unadjust"</span><span class="op">)</span>, nes_cutoff <span class="op">=</span> <span class="fl">1.5</span>, pval_cutoff <span class="op">=</span> <span class="fl">0.1</span>, pvalflag <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    interactive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Cell_type_deconvolution_and_differential_analysis_files/figure-html/unnamed-chunk-17-1.png" width="768" style="display: block; margin: auto;"></p>
<p>We can then generate summary plot of top enriched gene-sets.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gsea_sumplot.html">gsea_sumplot</a></span><span class="op">(</span>gseares_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/gsea/results/ct_adjust/"</span><span class="op">)</span>,</span>
<span>    pos_sel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GLYCOSPHINGOLIPID_BIOSYNTHESIS"</span>, <span class="st">"FRUCTOSE_AND_MANNOSE_METABOLISM"</span>, <span class="st">"HIF-1_SIGNALING_PATHWAY"</span><span class="op">)</span>,</span>
<span>    neg_sel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P53_SIGNALING_PATHWAY"</span>, <span class="st">"RENIN_SECRETION"</span>, <span class="st">"LYSOSOME"</span><span class="op">)</span>, pvalflag <span class="op">=</span> <span class="cn">FALSE</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Cell_type_deconvolution_and_differential_analysis_files/figure-html/unnamed-chunk-18-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Heatmap of two selected gene-sets for demonstration.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gsea_heatmap.html">gsea_heatmap</a></span><span class="op">(</span>normdata <span class="op">=</span> <span class="va">deres</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, teststats <span class="op">=</span> <span class="va">deres</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, gmtfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/gsea/gmtfile/kegg.gmt"</span><span class="op">)</span>, numgenes <span class="op">=</span> <span class="fl">200</span>, gsname_up <span class="op">=</span> <span class="st">"HIF-1_SIGNALING_PATHWAY"</span>,</span>
<span>    gsname_down <span class="op">=</span> <span class="st">"P53_SIGNALING_PATHWAY"</span>, anncol <span class="op">=</span> <span class="va">sampleinfo</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>,</span>
<span>        <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Cell_type_deconvolution_and_differential_analysis_files/figure-html/unnamed-chunk-19-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Random-walk plot of selected gene-sets</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gsea_rwplot.html">gsea_rwplot</a></span><span class="op">(</span>gseares_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCdeconR"</span><span class="op">)</span>, <span class="st">"/gsea/results/ct_adjust/"</span><span class="op">)</span>,</span>
<span>    gsname <span class="op">=</span> <span class="st">"HIF-1_SIGNALING_PATHWAY"</span>, class_name <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Cell_type_deconvolution_and_differential_analysis_files/figure-html/unnamed-chunk-20-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yuanhang Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
