<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Deconvolution of bulk RNA-seq data — scdecon • SCdeconR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Deconvolution of bulk RNA-seq data — scdecon"><meta property="og:description" content="Deconvolution of bulk RNA-seq data based on single-cell reference data. Eight bulk deconvolution methods, along with eight normalization
methods and four transformation methods are available."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCdeconR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Cell_type_deconvolution_and_differential_analysis.html">Cell type deconvolution and differential analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Liuy12/SCdeconR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Deconvolution of bulk RNA-seq data</h1>
    <small class="dont-index">Source: <a href="https://github.com/Liuy12/SCdeconR/blob/HEAD/R/deconvolution.R" class="external-link"><code>R/deconvolution.R</code></a></small>
    <div class="hidden name"><code>scdecon.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Deconvolution of bulk RNA-seq data based on single-cell reference data. Eight bulk deconvolution methods, along with eight normalization
methods and four transformation methods are available.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scdecon</span><span class="op">(</span></span>
<span>  <span class="va">bulk</span>,</span>
<span>  <span class="va">ref</span>,</span>
<span>  <span class="va">phenodata</span>,</span>
<span>  filter_ref <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  decon_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scaden"</span>, <span class="st">"CIBERSORT"</span>, <span class="st">"OLS"</span>, <span class="st">"nnls"</span>, <span class="st">"FARDEEP"</span>, <span class="st">"RLR"</span>, <span class="st">"MuSiC"</span>,</span>
<span>    <span class="st">"SCDC"</span><span class="op">)</span>,</span>
<span>  norm_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"LogNormalize"</span>, <span class="st">"TMM"</span>, <span class="st">"median_ratios"</span>, <span class="st">"TPM"</span>, <span class="st">"SCTransform"</span>,</span>
<span>    <span class="st">"scran"</span>, <span class="st">"scater"</span>, <span class="st">"Linnorm"</span><span class="op">)</span>,</span>
<span>  trans_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"log"</span>, <span class="st">"sqrt"</span>, <span class="st">"vst"</span><span class="op">)</span>,</span>
<span>  gene_length <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lfc_markers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span>,</span>
<span>  marker_strategy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"pos_fc"</span>, <span class="st">"top_50p_logFC"</span>, <span class="st">"top_50p_AveExpr"</span><span class="op">)</span>,</span>
<span>  to_remove <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ffpe_artifacts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cibersortpath <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pythonpath <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tmpdir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1234</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>bulk</dt>
<dd><p>a matrix or data.frame of unnormalizaed &amp; untransformed bulk RNA-seq gene expression values with rows representing genes, columns
representing samples</p></dd>


<dt>ref</dt>
<dd><p>a matrix or data.frame of untransformed scRNA-seq gene expression counts with rows representing genes, columns representing cells.
This data will be used to deconvolute provided bulk RNA-seq data.</p></dd>


<dt>phenodata</dt>
<dd><p>a data.frame with rows representing cells, columns representing cell attributes. It should at least contain the first three
columns as:</p><ol><li><p>cell barcodes</p></li>
<li><p>cell types</p></li>
<li><p>subject ids</p></li>
</ol></dd>


<dt>filter_ref</dt>
<dd><p>logical value indicating whether outlier genes &amp; cells should be removed from the provided reference data. Defaults to TRUE</p></dd>


<dt>decon_method</dt>
<dd><p>character value specifying the deconvolution method to use. Has to be one of "scaden", "CIBERSORT", "OLS", "nnls", "FARDEEP", "RLR",
"MuSiC", "SCDC". See details for more information.</p></dd>


<dt>norm_method</dt>
<dd><p>character value specifying the normalization method to use for both bulk &amp; reference data. Has to be one of "none","LogNormalize", "TMM",
"median_ratios", "TPM", "SCTransform", "scran", "scater", "Linnorm". See details for more information.</p></dd>


<dt>trans_method</dt>
<dd><p>character value specifying the transformation method to use for both bulk &amp; reference data. Has to be one of "none", "log", "sqrt",
"vst". See details for more information.</p></dd>


<dt>gene_length</dt>
<dd><p>a data.frame with two columns. The first column represents gene names that match with provided bulk data. The second column
represents length of each gene. Only applicable when norm_method is selected as "TPM".</p></dd>


<dt>lfc_markers</dt>
<dd><p>log2 fold change cutoff used to identify marker genes for deconvolution. The option only applicable to marker-gene based
approaches, such as CIBERSORT, OLS, nnls, FARDEEP and RLR.</p></dd>


<dt>marker_strategy</dt>
<dd><p>further strategy in selecting marker genes besides applying the log2 fold change cutoff. Can be chosen from: "all", "pos_fc",
"top_50p_logFC" or "top_50p_AveExpr". See details for more information.</p></dd>


<dt>to_remove</dt>
<dd><p>character value representing the cell type to remove from reference data. Only applicable to simulation experiments in evaluating
effect of cell type removal from reference.</p></dd>


<dt>ffpe_artifacts</dt>
<dd><p>logical value indicating whether to add simulated ffpe artifacts in the bulk data. Only applicable to simulation experiments in
evaluating the effect of FFPE artifacts.</p></dd>


<dt>prop</dt>
<dd><p>a matrix or data.frame of simulated cell proportion values with rows representing cell types, columns representing samples. Only applicable to simulation
experiments in evaluating the effect of cell type removal from reference.</p></dd>


<dt>cibersortpath</dt>
<dd><p>full path to CIBERSORT.R script.</p></dd>


<dt>pythonpath</dt>
<dd><p>full path to python binary where scaden was installed with.</p></dd>


<dt>tmpdir</dt>
<dd><p>temporary processing directory for scaden.</p></dd>


<dt>seed</dt>
<dd><p>random seed used for simulating FFPE artifacts. Only applicable when ffpe_artifacts is set to TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list containing two or three elements.</p><dl><dt>first element</dt>
<dd><p>a data.frame of predicted cell-type proportions, with rows representing cell types, columns representing samples.</p></dd>

 <dt>second element</dt>
<dd><p>a data.frame of fitting errors of the algorithm; first column represents sample names, second column represents RMSEs.</p></dd>

 <dt>optional third element</dt>
<dd><p>a data.frame of simulated cell proportion after removing the specified cell_type. Only applicable to simulation experiments.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>decon_method should be one of the following:</p><dl><dt><a href="https://github.com/KevinMenden/scaden" class="external-link">scaden</a></dt>
<dd><p>a deep learning based method using three multi-layer deep neural nets. To use scaden,
you need to firstly install scaden via <code>pip</code> or <code>conda</code>, the provide the python path to <code>pythonpath</code> option.</p></dd>

 <dt>CIBERSORT</dt>
<dd><p>a marker gene based support vectors regression approach. CIBERSOR does not allow redistribution. To use CIBERSORT,
you need to request the source code from the authors &amp; provide the path of CIBERSORT.R script to <code>cibersortpath</code> option.</p></dd>

 <dt>OLS</dt>
<dd><p>ordinary least squares.</p></dd>

 <dt><a href="https://cran.r-project.org/web/packages/nnls/index.html" class="external-link">nnls</a></dt>
<dd><p>non-negative least squares.</p></dd>

 <dt><a href="https://cran.r-project.org/web/packages/FARDEEP/index.html" class="external-link">FARDEEP</a></dt>
<dd><p>robust regression using least trimmed squares</p></dd>

 <dt><a href="https://cran.r-project.org/web/packages/MASS/index.html" class="external-link">RLR</a></dt>
<dd><p>robust regression using an M estimator</p></dd>

 <dt><a href="https://github.com/xuranw/MuSiC" class="external-link">MuSiC</a></dt>
<dd><p>multi-subject single-cell deconvolution</p></dd>

 <dt><a href="https://github.com/meichendong/SCDC" class="external-link">SCDC</a></dt>
<dd><p>an ENSEMBLE method to integrate deconvolution results from different scRNA-seq datasets</p></dd>


</dl><p>norm_method should be one of the following:</p><dl><dt>none</dt>
<dd><p>no normalization is performed.</p></dd>

 <dt>LogNormalize</dt>
<dd><p><code><a href="https://satijalab.org/seurat/reference/LogNormalize.html" class="external-link">LogNormalize</a></code> method from seurat.</p></dd>

 <dt>TMM</dt>
<dd><p>TMM method from <code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></code> function from edgeR.</p></dd>

 <dt>median_ratios</dt>
<dd><p>median ratio method from <code><a href="https://rdrr.io/pkg/DESeq2/man/estimateSizeFactors.html" class="external-link">estimateSizeFactors,DESeqDataSet-method</a></code> function from DESeq2.</p></dd>

 <dt>TPM</dt>
<dd><p>Transcript per million. TPM has to be chosen if ffpe_artifacts is set to TRUE.</p></dd>

 <dt>SCTransform</dt>
<dd><p><code>SCTransfrom</code> method from Seurat.</p></dd>

 <dt>scran</dt>
<dd><p><code><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></code> method from scran.</p></dd>

 <dt>scater</dt>
<dd><p><code><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html" class="external-link">librarySizeFactors</a></code> method from scater.</p></dd>

 <dt>Linnorm</dt>
<dd><p><code><a href="https://rdrr.io/pkg/Linnorm/man/Linnorm.html" class="external-link">Linnorm</a></code> method from Linnorm.</p></dd>


</dl><p>trans_method should be one of the following:</p><dl><dt>none</dt>
<dd><p>no transformation is performed.</p></dd>

 <dt>log2</dt>
<dd><p>log2 transformation. 0.1 is added to the data to avoid logarithm of 0s.</p></dd>

 <dt>sqrt</dt>
<dd><p>square root transformation.</p></dd>

 <dt>vst</dt>
<dd><p><code><a href="https://rdrr.io/pkg/DESeq2/man/varianceStabilizingTransformation.html" class="external-link">varianceStabilizingTransformation</a></code> method from DESeq2.</p></dd>


</dl><p>marker_strategy should be one of the following:</p><dl><dt>all</dt>
<dd><p>all genes passed fold change threshold will be used.</p></dd>

 <dt>pos_fc</dt>
<dd><p>only genes with positive fold changes will be used.</p></dd>

 <dt>top_50p_logFC</dt>
<dd><p>only genes with top 50 percent positive fold changes will be used.</p></dd>

 <dt>top_50p_AveExpr</dt>
<dd><p>only genes with top 50 percent average expression will be used.</p></dd>


</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co">## generate artificial bulk samples</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">refdata</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">phenodata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>cellid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">refdata</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                        celltypes <span class="op">=</span> <span class="va">refdata</span><span class="op">$</span><span class="va">celltype</span>,</span></span>
<span class="r-in"><span>                        subjectid <span class="op">=</span> <span class="va">refdata</span><span class="op">$</span><span class="va">subjectid</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bulk_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="bulk_generator.html">bulk_generator</a></span><span class="op">(</span>ref <span class="op">=</span> <span class="fu">GetAssayData</span><span class="op">(</span><span class="va">refdata</span>, slot <span class="op">=</span> <span class="st">"data"</span>, assay <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                           phenodata <span class="op">=</span> <span class="va">phenodata</span>, </span></span>
<span class="r-in"><span>                           num_mixtures <span class="op">=</span> <span class="fl">20</span>, </span></span>
<span class="r-in"><span>                           num_mixtures_sprop <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## perform deconvolution based on "OLS" algorithm</span></span></span>
<span class="r-in"><span><span class="va">decon_res</span> <span class="op">&lt;-</span> <span class="fu">scdecon</span><span class="op">(</span>bulk <span class="op">=</span> <span class="va">bulk_sim</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                     ref <span class="op">=</span> <span class="fu">GetAssayData</span><span class="op">(</span><span class="va">refdata</span>, slot <span class="op">=</span> <span class="st">"data"</span>, assay <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     phenodata <span class="op">=</span> <span class="va">phenodata</span>,</span></span>
<span class="r-in"><span>                     filter_ref <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                     decon_method <span class="op">=</span> <span class="st">"OLS"</span>,</span></span>
<span class="r-in"><span>                     norm_method <span class="op">=</span> <span class="st">"none"</span>,</span></span>
<span class="r-in"><span>                     trans_method <span class="op">=</span> <span class="st">"none"</span>,</span></span>
<span class="r-in"><span>                     marker_strategy <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yuanhang Liu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

